### PR TITLE
Thread safe ISBRH compatibility with Extra Utilities 2

### PR DESCRIPTION
**Pull request by Cleptomania**
Since GTNH does not have a fork for Extra Utilities 2, I have implemented the thread safety fixes for it directly in Angelica via a combination of ASM/mixins. 

This does lay a small bit of ground work for potentially other mods to use the ASM transformer I've put in place, if we should need to do these types of fixes for other mods we don't control the source of.

This largely fixes https://github.com/GTNewHorizons/Angelica/issues/425, which was caused by EnderIO being marked as thread-safe for painted blocks, but Extra Utilities was not. However that may be a larger issue such that anything drawn by EnderIO painted blocks has to be thread-safe, but that's more of a concern for EnderIO than Angelica directly. Extra Utilities was the first one that has been reported like this.

I've tested most of the blocks this effects both with the stand-alone EU2 blocks and also via EnderIO painting.

### CODE REVIEW

File: src/main/java/com/gtnewhorizons/angelica/mixins/Mixins.java [hunk 1]
```diff
@@ -183,6 +184,14 @@ public enum Mixins {
             "angelica.animation.MixinWorldRenderer",
             "angelica.animation.MixinRenderItem")),
 
+    EXTRA_UTILITIES_THREAD_SAFETY(new Builder("Enable thread safety fixes in Extra Utilities 2").setPhase(Phase.LATE)
```
Comment (mitchej123) [inline src/main/java/com/gtnewhorizons/angelica/mixins/Mixins.java:187]: nit: I Think it's only exu on 1.7.10 and exu2 on 1.12?

File: src/main/java/com/gtnewhorizons/angelica/config/AngelicaConfig.java [hunk 2]
```diff
@@ -101,6 +101,11 @@ public class AngelicaConfig {
     @Config.RequiresMcRestart
     public static boolean fixLotrSodiumCompat;
 
+    @Config.Comment("Fix thread-safety in Extra Utilities 2 rendering")
+    @Config.DefaultBoolean(true)
+    @Config.RequiresMcRestart
+    public static boolean fixEU2SodiumCompat;
```
Comment (mitchej123) [inline src/main/java/com/gtnewhorizons/angelica/config/AngelicaConfig.java:107]: just EU

File: src/main/java/com/gtnewhorizons/angelica/mixins/late/client/extrautils/MixinIconConnectedTexture.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import org.spongepowered.asm.mixin.Overwrite;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.Unique;
+
+@Mixin(IconConnectedTexture.class)
+public class MixinIconConnectedTexture {
+
+    @Final
+    @Shadow(remap = false)
+    public IIcon[] icons;
```
Comment (mitchej123) [inline src/main/java/com/gtnewhorizons/angelica/mixins/late/client/extrautils/MixinIconConnectedTexture.java:18]: Is this going to cause any threading troubles?

File: src/main/java/com/gtnewhorizons/angelica/transform/compat/extrautils/RenderBlockColorTransformer.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import org.objectweb.asm.tree.AnnotationNode;
+import org.objectweb.asm.tree.ClassNode;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+
+public class RenderBlockColorTransformer implements IClassTransformer {
+
+    @Override
+    public byte[] transform(String name, String transformedName, byte[] basicClass) {
```
Comment (mitchej123) [inline src/main/java/com/gtnewhorizons/angelica/transform/compat/extrautils/RenderBlockColorTransformer.java:18]: NIT: Might be better to do one transformer with a map of the classes and perThread true/false for each?

Comment (mitchej123) [reply src/main/java/com/gtnewhorizons/angelica/transform/compat/extrautils/RenderBlockColorTransformer.java:18]: Also might be worth using the class constant parser to ensure this doesn't blindly try to transform every class and slow startup down.

**Review by mitchej123 — COMMENTED** [review]
Nice, thanks!  The transformers feel as if they could be merged and the class constant parser would be a nice addition.  Will approve after you've looked through and responded to the comments.

File: src/main/java/com/gtnewhorizons/angelica/transform/compat/extrautils/RenderBlockColorTransformer.java (hunk 4 above)
Comment (Cleptomania) [reply src/main/java/com/gtnewhorizons/angelica/transform/compat/extrautils/RenderBlockColorTransformer.java]: I believe this is handled with the latest commit, it's not using class constant parser but I think achieves the same goal

File: src/main/java/com/gtnewhorizons/angelica/transform/compat/ExtraUtilsTransformer.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        perThreadMap.put(RenderBlockConnectedTexturesEthereal, true);
+        perThreadMap.put(RenderBlockFullBright, false);
+        perThreadMap.put(RenderBlockSpike, false);
+    }
+
+    @Override
+    public byte[] transform(final String className, String transformedName, byte[] basicClass) {
+        if (basicClass == null) return null;
+
+        if (!transformedClasses.contains(transformedName)) {
```
Comment (mitchej123) [inline src/main/java/com/gtnewhorizons/angelica/transform/compat/ExtraUtilsTransformer.java:41]: Might be worth profiling this vs the CST parser, but as long as load time in the full pack isn't too bad it's probably fine to test as is for now

**Pull request merged**